backend/init/init.go

```diff
@@ -73,3 +74,36 @@ func Set(name string, f func() backend.Backend) {
 
 	backends[name] = f
 }
+
+// deprecatedBackendShim is used to wrap a backend and inject a deprecation
+// warning into the Validate method.
+type deprecatedBackendShim struct {
+	backend.Backend
+	Message string
+}
+
+// Validate the Backend then add the deprecation warning.
+func (b deprecatedBackendShim) Validate(c *terraform.ResourceConfig) ([]string, []error) {
+	warns, errs := b.Backend.Validate(c)
+	warns = append(warns, b.Message)
+	return warns, errs
+}
+
+// DeprecateBackend can be used to wrap a backend to retrun a deprecation
+// warning during validation.
+func deprecateBackend(b backend.Backend, message string) func() backend.Backend {
+	if _, ok := b.(backend.Enhanced); ok {
+		panic("cannot use DeprecateBackend on an Enhanced Backend")
+	}
+
+	if _, ok := b.(backend.Local); ok {
+		panic("cannot use DeprecateBackend on a Local Backend")
+	}
```

Comment:
It's because when you embed an interface, you lose the ability to assert any other optional interfaces later on (this may have been what you were saying already). In order to handle those cases this would need to separate implement all the combinations, i.e. `BackendShim`, `EnhancedBackendShim`, `LocalBackendShim`, and `EnhcancedLocalBackendShim` and return the correct type. 

I'll add a little more clarification, but I think once a more advanced schema generated backend needs to be deprecated, we'll add that ability through the schema rather than here.